src/macros.rs

```diff
@@ -37,34 +39,65 @@ macro_rules! redis_command {
             )
         } == $crate::raw::Status::Err as c_int
         {
+            $crate::raw::redis_log(
+                $ctx,
+                &format!("Error: failed to create command {}", $command_name),
+            );
             return $crate::raw::Status::Err as c_int;
         }
 
-        if $acl_categories != "" {
-            let acl_categories = CString::new($acl_categories).unwrap();
+        let command =
+            unsafe { $crate::raw::RedisModule_GetCommand.unwrap()($ctx, name.as_ptr()) };
+        if command.is_null() {
+            $crate::raw::redis_log(
+                $ctx,
+                &format!("Error: failed to get command {}", $command_name),
+            );
+            return $crate::raw::Status::Err as c_int;
+        }
 
-            let command =
-                unsafe { $crate::raw::RedisModule_GetCommand.unwrap()($ctx, name.as_ptr()) };
-            if command.is_null() {
-                return $crate::raw::Status::Err as c_int;
-            }
+        if let Some(RM_SetCommandACLCategories) = $crate::raw::RedisModule_SetCommandACLCategories {
+            let mut acl_categories = CString::new("").unwrap();
+            $(
+                if $mandatory_acl_categories != "" && $optional_acl_categories != "" {
+                    acl_categories = CString::new(format!("{} {}", $mandatory_acl_categories, $optional_acl_categories)).unwrap();
+                } else if $optional_acl_categories != "" {
+                    acl_categories = CString::new($optional_acl_categories).unwrap();
+                }
+                // Warn if optional ACL categories are not set, but don't fail.
+                if RM_SetCommandACLCategories(command, acl_categories.as_ptr()) == $crate::raw::Status::Err as c_int {
+                    $crate::raw::redis_log(
+                        $ctx,
+                        &format!(
+                            "Warning: failed to set command `{}` ACL categories `{}`",
+                            $command_name, acl_categories.to_str().unwrap()
+                        ),
+                    );
+                } else
+            )?
+            if $mandatory_acl_categories != "" {
+                acl_categories = CString::new($mandatory_acl_categories).unwrap();
 
-            if let Some(RM_SetCommandACLCategories) =
-                $crate::raw::RedisModule_SetCommandACLCategories
-            {
+                // Fail if mandatory ACL categories are not set.
                 if RM_SetCommandACLCategories(command, acl_categories.as_ptr())
                     == $crate::raw::Status::Err as c_int
                 {
                     $crate::raw::redis_log(
                         $ctx,
                         &format!(
-                            "Error: failed to set command {} ACL categories {}",
-                            $command_name, $acl_categories
+                            "Error: failed to set command `{}` mandatory ACL categories `{}`",
```

Comment:
I have forgotten; don't the macro arguments work well with the format arguments just yet, so can't they be inlined? I think the dollar sign breaks the parser there.

Reply:
the `$` does break the parser, they can't be inlined (as macro args)

Reply:
:-( Thanks